### ISSUE TITLE
LwIP2 + SSL crash on remote server disconnect

### ISSUE THREAD

**Issue by Adam5Wu** [me-no-dev/ESPAsyncTCP#83]
When combining LwIP2 and SSL using TCP client, there seems to be a race condition when remote server disconnects.

According to my limited tests, crash do NOT happen with:
- LwIP2 + non-SSL
- LwIP + SSL
- LwIP2 + SSL + enable serial debug (ets_printf)

But with LwIP2 + SSL + disable serial debug, there is a high probability of crash (>80%) shortly after remote server initiated disconnect:
- If discard_cb is assigned, the crash happen after the callback completes;
- Same crash happens even when discard_cb is not assigned.

The crash stack trace looks like the following:
```
Exception 3: LoadStoreError: Processor internal physical address or data error during load or store
Decoding 27 results
0x4010011d: umm_assimilate_up at D:\Dev\arduino-1.8.0\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc/umm_malloc.c line 1167
0x401034c7: lmacRecycleMPDU at ?? line ?
0x4010392a: lmacRecycleMPDU at ?? line ?
0x4010020c: _umm_free at D:\Dev\arduino-1.8.0\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc/umm_malloc.c line 1291
0x40100690: free at D:\Dev\arduino-1.8.0\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc/umm_malloc.c line 1739
0x4021a2b0: mem_free at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/core/mem.c line 152
0x4021343a: memp_free at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/core/memp.c line 231
0x40102136: wDev_ProcessFiq at ?? line ?
0x40215e10: pbuf_free_LWIP2 at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/core/pbuf.c line 1306
0x40203ffd: tcp_ssl_read at D:\Sketches\libraries\ESPAsyncTCP\src/tcp_axtls.c line 477
0x40104e59: ets_timer_disarm at ?? line ?
0x401006fa: cont_run at D:\Dev\arduino-1.8.0\hardware\esp8266com\esp8266\cores\esp8266/cont.S line 89
0x40209bc3: AsyncClient::_recv(tcp_pcb*, pbuf*, long) at D:\Sketches\libraries\ESPAsyncTCP\src/ESPAsyncTCP.cpp line 665
0x40209bec: AsyncClient::_s_recv(void*, tcp_pcb*, pbuf*, long) at D:\Sketches\libraries\ESPAsyncTCP\src/ESPAsyncTCP.cpp line 665
0x40213ba4: tcp_process_refused_data at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/core/tcp.c line 1779
0x40215766: tcp_input at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/core/tcp_in.c line 357
0x40219a1d: ip4_input at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/core/ipv4/ip4.c line 685
0x402132e5: ethernet_input_LWIP2 at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/lwip2-src/src/netif/ethernet.c line 182
0x4010453c: esp2glue_ethernet_input at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/glue-lwip/lwip-git.c line 406
0x4023bbd4: ethernet_input at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/glue-esp/lwip-esp.c line 356
0x4023bbe6: ethernet_input at /local/users/gauchard/arduino/arduino_esp8266/origin/tools/sdk/lwip2/builder/glue-esp/lwip-esp.c line 364
0x40236c0f: ppPeocessRxPktHdr at ?? line ?
0x4023dbef: ets_snprintf at ?? line ?
```

**Comment by Adam5Wu** [comment]
LOL turned out it is not a race condition but a double-free.
The undeterminism mainly came from uninitialized memory status, and turning on debug perturbed the address space enough to hide the problem.

**Comment by Adam5Wu** [comment]
Fix proposed in PR #84
